File: _book/source/_posts/Algorithm/1.leetcode/8.字符串转换整数/8.字符串转换整数-atoi.py
#
# @lc app=leetcode.cn id=8 lang=python3
#
# [8] 字符串转换整数 (atoi)
#

# @lc code=start
class Solution:
    def myAtoi(self, str: str) -> int:
        """
        :type str: str
        :rtype: int
        """
        # 1.去掉前后空格
        str = str.strip()
        if not str:
            return 0
        MAX_INT = 2147483647
        MIN_INT = -2147483648
        ret:int =0
        pos:int = 0;sign:int =1;overflow =False
        # 取正负符号，如果存在，位置加1
        if str[pos]=='-':
            pos+=1
            sign = -1
        elif str[pos] =='+':
            pos += 1
        # 从pos 算起， range()  ,如果是不是数据，返回， 如果计算后的结果，超过正常范围则 溢出返回 最大、或最小
        for i in range(pos,len(str)):
            if not str[i].isdigit():
                break
            ret = ret *10 + int(str[i])
            if not MIN_INT<= sign * ret <=MAX_INT:
                overflow = True
                break
        if overflow:
            return MAX_INT if sign==1 else MIN_INT
        else:
            return sign*ret
# @lc code=end

File: _book/source/_posts/Algorithm/1.leetcode/8.字符串转换整数/8.字符串转换整数-atoi.go
/*
 * @lc app=leetcode.cn id=8 lang=golang
 *
 * [8] 字符串转换整数 (atoi)
 */
package main

import (
	"fmt"
	"math"
	"strings"
)

// @lc code=start
func myAtoi(str string) int {
	strTrim := strings.TrimLeft(str, " ")
	if len(strTrim) == 0 {
		return 0
	}
	sign := 1
	result := 0
	i := 0
	if strTrim[0] == '-' {
		sign = -1
		i = 1
	} else if strTrim[0] == '+' {
		sign = 1
		i = 1
	}
	for ; i < len(strTrim); i++ {
		ch := int(strTrim[i] - '0')
		tem := result*10 + ch
		//判断char 是否是数字
		if ch < 0 || ch > 9 {
			break
		} else if sign == 1 && tem > math.MaxInt32 {
			return math.MaxInt32
		} else if sign == -1 && tem*sign < math.MinInt32 {
			return math.MinInt32
		}
		result = tem
	}
	return result * sign
}

// @lc code=end

func main() {
	fmt.Println("start")
	// fmt.Println(myAtoi("  -42"))
	fmt.Println(myAtoi("4193 with words"))
}
